### DISCUSSION TITLE
Possible to be in only in some parallel states?

### DISCUSSION THREAD

**Discussion by lougreenwood** [statelyai/xstate#1454] [General] [answered]
I have a state machine which is used to drive the state which a button component can be in and by inference, the action(s) the button should perform.

Sometimes the button can allow multiple actions, in this case, the button has a dropdown component which displays secondary actions.

This means that when the state machine is in specific states, it should allow an arbitrary number of sub-states to be enabled.

These multiple actions are modelled in the state chart as sibling parallel states:

```js
// "create" is a state which will be rendered as multiple actions
createMultipleThings: {
  type: 'parallel',
  states: {
    createThingA: {},
    createThingB: {},
    createThingC: {},
  }
}
```

The problem I'm facing is that in some contexts, the button should only show a subset of these states. e.g `Page Foo` may not allow `thingB` to be created, so the state machine should only have a state of `['createMultipleThings.createThingA', 'createMultipleThings.createThingC]`.

However, it seems that parallel states require that the machine is in all states at once, but not some of all.

It seems that the way I should model my button is that each of the parallel states should have an `enabled/disabled` sub-state to represent what actions are available, but that is counter to the design of the rest of my state machine, which uses the leaf-most node in the state chart to represent what action is possible, e.g:

- `actions.deleteThingXYZ`
  - ✅ button should perform `deleteThingXYZ` operation

- `createMultipleThings.createThingA.enabled`
  - ❌ There's no action called `enabled`, but there is a `createThingA` action

So my question is, is it possible to be in just some, but not all `state`s in a parallel `stateNode`?

**Comment by rjdestigter** [comment]
I have a feeling you could be taking a different approach to the design of the state machine but I'm having trouble understanding the requirements here. Could you just describe the button requirements?

**Reply by lougreenwood** [reply]
> Could you just describe the button requirements?

Sure, I'll do my best 😃.

The button component should receive a running state machine and from that state machine, the button should derive it's own configuration from the current state.

From a user's perspective, the button defines the actions available to a user at a particular point in a workflow.

Each top-level state in the state machine represents a "milestone" in this workflow. If that milestone allows multiple actions, they are modelled as parallel child states.

Completing an action may progress a user through the workflow by modifying the model associated with the workflow and having the Statechart react to the updated model values.

The button can either perform one action (rendered as just a simple button component), or many actions (a segmented button which allows triggering a primary actions on click of the button and selecting secondary actions by opening a dropdown via the button segment.)

The button will determine it's configuration based upon the presence of a specific state (aka "workflow milestone") being active.

Some examples of the button configuration:

<img width="198" alt="Screenshot 2020-09-12 at 15 48 24" src="https://user-images.githubusercontent.com/12051736/92998102-d7ad6100-f50f-11ea-8911-32282634e4f7.png">
<img width="191" alt="Screenshot 2020-09-12 at 15 48 32" src="https://user-images.githubusercontent.com/12051736/92998104-d8de8e00-f50f-11ea-9aa1-8e4cb1e9ae9c.png">
<img width="181" alt="Screenshot 2020-09-12 at 15 52 22" src="https://user-images.githubusercontent.com/12051736/92998127-f875b680-f50f-11ea-8314-242eb5eca227.png">

The state machine handles detection of whether a given state can be entered based upon various logic, such as whether a model (in `context`) has a specific feature.

The state machine (currently) models a segmented dropdown as being in a specific parallel state where each child state represents one of the button actions (either the primary feature (e.g `Create Thing A`), or the dropdown actions) - this is where I'm hitting a problem. It seems that state charts don't allow being in some but not all states of a parallel parent. It seems like an obvious permutation of the types of behaviour a state chart is designed to model, so I can only assume there's a good reason and an alternate pattern to follow?

<details>
  <summary>Here's my current state machine (which works perfectly, except for not being able to optionally be in only some parallel state 🙁</summary>

```js
const CRM_ACTIONS_TO_SUPPORTED_FEATURES_MAP = {
  connect: 'cfaSendToCrmAction',
  confirmMatch: {
    withOrganization: 'organization',
    withLead: 'crmLeads',
  },
  connectLead: {
    asOrganization: 'organization',
    asLead: 'crmLeads',
  },
  createCrmEntity: {
    opportunity: 'opportunities',
    task: 'tasks',
  },
};

const GUARDED_TRANSITIONS = {
  confirmMatch: {
    target: '#actions.connected.confirmMatch',
    cond: 'shouldTranstionToConfirmMatch',
  },
  connectLead: {
    target: '#actions.connected.connectLead',
    cond: 'shouldTransitionToConnectLead',
  },
  connected: { target: '#actions.connected', cond: 'isConnected' },
  createCrmEntity: {
    target: '#actions.connected.createCrmEntity',
    cond: 'shouldTransitionToCreateCrmEntity',
  },
};

// The order of this is very important and defines the order in which we will define transitions through the machine.
const guardedTransitionsAfterConnectInOrder = [
  GUARDED_TRANSITIONS.confirmMatch,
  GUARDED_TRANSITIONS.connectLead,
  GUARDED_TRANSITIONS.createCrmEntity,
];

export default Machine(
  {
    id: 'crmWorkflow',
    type: 'parallel',
    context: {
      lead: {
        crmIntegrationConnection: {
          integartion: { id: 1 },
        },
        hasMatches: false,
        sentToCrm: false,
      },
      account: {
        crmIntegrationConnection: {
          integration: { id: 1 },
        },
      },
    },
    states: {
      compatibility: {
        initial: 'unknown',
        states: {
          unknown: {
            on: {
              '': [{ target: 'compatible', cond: { type: 'isCompatibleCrm' } }, { target: 'incompatible' }],
            },
          },
          incompatible: {},
          compatible: {
            entry: send('CRM_IS_COMPATIBLE'),
            on: {
              CRM_IS_COMPATIBLE: '#crmWorkflow.actions',
            },
          },
        },
      },
      actions: {
        id: 'actions',
        initial: 'unknown',
        states: {
          unknown: {
            on: {
              '': [GUARDED_TRANSITIONS.connected, { target: '#actions.disconnected' }],
            },
          },
          disconnected: {
            on: {
              CRM_CONNECTED: 'connected',
            },
          },
          connected: {
            initial: 'unknown',
            on: {
              NEXT: {
                target: [...guardedTransitionsAfterConnectInOrder, { target: '#actions.connected.done' }],
              },
            },
            states: {
              unknown: {
                on: {
                  '': [...guardedTransitionsAfterConnectInOrder, { target: 'done' }],
                },
              },

              confirmMatch: {
                '': [{ target: 'one', cond: 'hasOneMatch' }, { target: 'many' }],
                states: {
                  one: {},
                  many: {},
                },
              },

              connectLead: {
                type: 'parallel',
                // TODO: check if supports either of these features and transition into one/many
                states: {
                  asOrganization: {},
                  asLead: {},
                },
              },

              createCrmEntity: {
                type: 'parallel',
                // TODO: check if supports either of these features and transition into one/many
                states: {
                  opportunity: {},
                  task: {},
                },
              },

              // State to transition into if no more actions are available
              done: {
                type: 'final'
              },
            },
          },
        },
      },
    },
  },
}
```
</details>

**Comment by Andarist** [comment]
To answer the original question - not, it's not possible. All parallel states are always entered. 

You need to either model some of this using compound states so you could make things mutually exclusive or introduce disabled/enabled states into your parallel regions.

**Reply by lougreenwood** [reply]
That's a shame - it means I need to add complexity to a simple design or move validation of the parallel states to the consumer - i.e the button component.

As someone just getting familiar with state charts, it seems that being able to enter "some of many" sibling/parallel states would be a fundamental tool for modelling state, it seems like such a large and obvious hole in the language of state modelling that it can't be accidental?

So I'm wondering - again, as a newbie...  is there some fundamental property of state charts that means this "some of many" design isn't allowed?
Am I missing some key piece of information about state chart basics that would help me explain to someone that "this is why we can't model being in "some" of many sibling states, it's really obvious, it's because you have to think of state charts as..."

**Reply by rjdestigter** [reply]
I still think you should be approaching this differently but I'm still confused on the actual requirements. Could you just talk about the requirements from a users perspective? Forget about state machines, don't mention them. Forget about trying to make some generic button component that works for several workflows. Just pick one scenario and describe in a few lines what the user should experience.

**Reply by lougreenwood** [reply]
The user is on a sales lead page in the main application.
The sales lead is associated with an external CRM software.

The page displays a button which presents the user with a set of contextual actions which the user can perform on the lead in the external CRM software. The choice of contextual action is limited by:
- which CRM the user has connected to the lead
- the feature set which that specific CRM supports
- previous actions which have been performed on the lead on the external CRM and/or in the main application.

The specific contextual actions are designed to assist the user to increment the progress of the sales lead through a sales funnel or better integrate the lead into the external CRM.

**Reply by mattpocock** [reply]
Could each substate (member of the parallel state) contain a 'disabled' state, in which it isn't shown?

Then each substate could calculate whether or not to disable itself.

**Reply by lougreenwood** [reply]
@mattpocock Thanks for the suggestion - sure, that's possible and will likely end up being my solution (unless I just decide to cut my losses and chose to make the state machine consumer responsible for checking the parallel states validity...).

I was hoping to avoid adding an extra set of nested states to indicate whether some parallel feature is enabled or not as it adds a whole bunch of extra complexity to what should be a simple machine. The compromises I see are:

- it adds (seemingly un-necessary) complexity to the state machine
- it leaks implementation details out into the state machine's interface (the existence of the "disabled" state is driven by restrictions  imposed by the state chart spec or whatever)
- it makes the state chart design inconsistent/compromises the API or requires adding even more complexity
  - only parallel states need this extra nested "disabled" state, but consumers of the state machine shouldn't care about these design constraints, so to make the API consistent, I should probably also add some "disabled" nested state to the compound states too, even though they have no need for it at all. And now my state machine as ballooned in size, complexity and knowledge required to understand it.

Which brings me back to my meta question - is there some fundamental reason that state charts don't allow modelling "some of many" states? Am I missing a key insight which will make all of this make sense, because I'd love to understand this better ☺️.

**Reply by Andarist** [reply]
> Which brings me back to my meta question - is there some fundamental reason that state charts don't allow modelling "some of many" states? Am I missing a key insight which will make all of this make sense, because I'd love to understand this better ☺️.

You'd have to denote this fact somehow on the statechart and given that you can build something like this using the existing primitives (extra internal disabled state) there probably wasn't much need to introduce more complexity to the syntax / visual representation.

You could not expose the raw API and check the complexity of "disabled" states within your custom thin layer, that way your consumer wouldn't have to worry about this nor your compound states would have to introduce disabled states.

**Comment by johnyanarella** [comment, marked as answer]
One hurdle I encountered as a newcomer to statecharts was overcoming how overloaded the meaning of the word "state" was in my head from past experience. It's easy to slip into thinking of modeling state in terms of modeling the shape of the state (ex. "some of many" above), rather than as states of being and transitions between those states.

A parallel state is divided into regions that are essentially separate nested state machines. Each of those has its own state of being. If you determine this is the best way to model your actions, each of those nested action state machines needs to capture all the states a given action can be in, and (as suggested above) it sounds like the missing one is "disabled" or "invisible".

A few other things that may or may not help unlock a path to something you find elegant:

- it might help to consider a synonym like "modes" in your head when thinking about the states in a state machine;
- identifying what "things" have "modes", and what those "modes" are, can help illuminate the boundaries of what should be modeled as separate machines—either via parallel states to allow visibility into shared state, or as separate actors;
- sometimes those "modes" aren't really finite, are not known up front, are very numerous, or are highly dynamic—this is where the context becomes useful; and
- the phrase "deriving its configuration" (aka deriving its state) makes me wonder if there's another "thing" here that goes through some "modes" of its own.

Statecharts are purpose built for representing situations where there are many states of being, operations vary by meaning or availability by state, and there are constrained transitions between those states. They are a versatile tool in your tool belt, but they're also not a golden hammer. You don't necessarily have to model the whole problem in one. It's okay to be practical about their use and pull in other data structures when they're well suited to address a part of the problem.

You know your particular problem domain best, but these things came to mind:

- it might make sense to just maintain an Array or Set of available button actions in the context and update those as you transition (via an entry action) through the states of your workflow machine;
- this is especially true if those actions vary wildly (are retrieved from a database, based on permissions, are going to grow numerous and change over time as you maintain this app, etc.);
- your button could be a statechart of its own (your example screenshots above implied it has a couple states based on the number and presence of particular actions) or its also okay to just handle that in your render logic until greater complexity demands a statechart.

**Comment by lucywang000** [comment]
The answer given by @ johnyanarella is excellent.

*(Coming across this thread by searching around "parallel states" keyword in this project, since I'm going to implement the parallel feature in [my own statecharts library](https://github.com/lucywang000/clj-statecharts).)*

Your requirement is to define a list of available actions that could be performed by the user on the current lead. The available actions is decided by a combination factors of `features supported by external CRM` plus `current stage of the lead`. 

If the above statement is correct, you are kinda "abusing" the parallel states feature here.

> Each top-level state in the state machine represents a "milestone" in this workflow. 

This is great and the idea use case of statecharts or state machines.

> If that milestone allows multiple actions, they are modelled as parallel child states.

This is the part that becomes a problem IMHO. In statecharts parallel states are created a particular use case, i.e. for containing multiple loosely coupled regions/parts within the same statecharts. This means each part is always there.

But  in your case, possible actions becomes parallel states, and IIUC you would like to "hide" some parallel states based on the features of the CRM software used, which is a quite "novel" usage pattern, and no doubt it's not well supported in the SCXML spec or xstate.

I would strongly suggest, if possible, adopt the suggested work around of @johnyanarella :
> it might make sense to just maintain an Array or Set of available button actions in the context and update those as you transition (via an entry action) through the states of your workflow machine;


Just my cents, YMMV :)
